test/test_datapipe.py

```diff
@@ -2703,5 +2706,59 @@ def reset(self):
             list(it)
         self.assertEqual(3, datapipe._number_of_samples_yielded)
 
+
+class TestIterDataPipeGraphFastForward(TestCase):
+
+    def _fast_forward_graph_test_helper(self, datapipe, fast_forward_fn, n_iterations=3, rng=None):
+        if rng is None:
+            rng = torch.Generator()
+        initial_rng_state = rng.get_state()
+        torch.utils.data.graph_settings.apply_shuffle_seed(datapipe, rng)
+        expected_full_res = list(datapipe)
+
+        # Test Case: fast forward works with list
+        rng.set_state(initial_rng_state)
+        fast_forward_fn(datapipe, n_iterations, rng)
+        actual_res = list(datapipe)
+        self.assertEqual(len(datapipe) - n_iterations, len(actual_res))
+        self.assertEqual(expected_full_res[n_iterations:], actual_res)
+
+        # Test Case: fast forward works with iterator
+        rng.set_state(initial_rng_state)
+        fast_forward_fn(datapipe, n_iterations, rng)
+        it = iter(datapipe)
+        actual_res = list(it)
+        self.assertEqual(len(datapipe) - n_iterations, len(actual_res))
+        self.assertEqual(expected_full_res[n_iterations:], actual_res)
+        with self.assertRaises(StopIteration):
+            next(it)
+
+    def test_simple_fast_forward_graph(self):
+        graph1 = dp.iter.IterableWrapper(range(10))
+        self._fast_forward_graph_test_helper(graph1, simple_fast_forward_graph)
+
+        graph2 = graph1.map(_mul_10)
+        self._fast_forward_graph_test_helper(graph2, simple_fast_forward_graph)
+
+        rng = torch.Generator()
+        graph3 = graph2.shuffle()
+        self._fast_forward_graph_test_helper(graph3, simple_fast_forward_graph, rng=rng)
+
+        graph4 = graph3.map(_mul_10)
+        self._fast_forward_graph_test_helper(graph4, simple_fast_forward_graph, rng=rng)
+
+        graph5 = graph4.batch(2)
+        self._fast_forward_graph_test_helper(graph5, simple_fast_forward_graph, rng=rng)
+
+        # With `fork` and `zip`
+        cdp1, cdp2 = graph5.fork(2)
+        graph6 = cdp1.zip(cdp2)
+        self._fast_forward_graph_test_helper(graph6, simple_fast_forward_graph, rng=rng)
+
+        # With `fork` and `concat`
+        graph7 = cdp1.concat(cdp2)
+        self._fast_forward_graph_test_helper(graph7, simple_fast_forward_graph, rng=rng)
+
+
```

Comment:
Missing tests:

What will happen if I get state, restore it on the new graph, do extra 5 iterations, save state and restore it on the new graph.

Reply:
This is in the next PR.

Reply:
I combined that PR with this one.